### PR TITLE
feat(Service Categories): Add Action Logs

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to:
- Add Action Logs while Create/Delete a service category through APIv2
- Add Action Logs on all the legacy part (was never been implemetend)
- Refactor on existing and already fixed action logs to remove magic strings in legacy

**Fixes** # MON-151475

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x
- [x] 24.04.x
- [x] 24.10.x
- [x] master

<h2> How this pull request can be tested ? </h2>

Create an SC using following payload:

POST /configuration/services/categories
```JSON
{
  "name": "service-category",
  "alias": "service-category",
  "is_activated": false,
}
```
on Database execute the query
```SQL
SELECT * FROM centreon_storage.log_action WHERE object_name = "service-category"
```
You should have the following result:

```
+---------------+-----------------+--------------+-----------+---------------------+-------------+----------------+
| action_log_id | action_log_date | object_type  | object_id | object_name         | action_type | log_contact_id |
+---------------+-----------------+--------------+-----------+---------------------+-------------+----------------+
|            18 |      1730801774 | servicecategories |         8 | service-category       | a           |              1 |
```

Delete the SC
DELETE /configuration/services/categories/<hc_id>

Play same SQL Request
You should have the following result:
```
+---------------+-----------------+--------------+-----------+---------------+-------------+----------------+
| action_log_id | action_log_date | object_type  | object_id | object_name   | action_type | log_contact_id |
+---------------+-----------------+--------------+-----------+---------------+-------------+----------------+
|            17 |      1730801738 | servicecategories |         1 | service-categpry    | d           |              1 |
+---------------+-----------------+--------------+-----------+---------------+-------------+----------------+
```

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by dmyios — DISMISSED** [review]

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbWriteHostTemplateActionLogRepository.php [hunk 1]
```diff
@@ -100,15 +100,15 @@ public function add(NewHostTemplate $hostTemplate): int
             if ($hostTemplateId === 0) {
                 throw new RepositoryException('Host template ID cannot be 0');
             }
-            
+
             $actionLog = new ActionLog(
-                self::HOST_TEMPLATE_OBJECT_TYPE,
+                ActionLog::OBJECT_TYPE_HOST,
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbWriteHostTemplateActionLogRepository.php:105]: Maybe use ActionLog::OBJECT_TYPE_HOST_TEMPLATE instead.

Comment (jeremyjaouen) [reply centreon/src/Core/HostTemplate/Infrastructure/Repository/DbWriteHostTemplateActionLogRepository.php:105]: Was thinking about it, but in legacy Host Templates are referenced as "host" in action logs and I didn't want to change the original behaviour.

I could also create a OBJECT_TYPE_HOST_TEMPLATE constant that will have "host" as value so original behaviour will still not be changed and be less confusing while reading.

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 2]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+                }
             }
         }
     }
     CentreonACL::duplicateScAcl($scAcl);
     $centreon->user->access->updateACL();
 }
 
-function enableServiceCategorieInDB(?int $sc_id = null, $sc_arr = [])
+function enableServiceCategorieInDB(?int $serviceCategoryId = null, $serviceCategories = [])
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:213]: You can use array no ? 
```suggestion
function enableServiceCategorieInDB(?int $serviceCategoryId = null, array $serviceCategories = [])
```

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 3]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
 {
-    if (!$sc_id && !count($sc_arr)) {
+
+    if (! $serviceCategoryId && ! count($serviceCategories)) {
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:216]: Empty is fatser than count :) 
```suggestion
    if (! $serviceCategoryId && empty($serviceCategories)) {
```

**Review by dmyios — APPROVED** [review]

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 4]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+            WHERE `sc_id` = :serviceCategoryId LIMIT 1
+        SQL
+    );
+    foreach (array_keys($serviceCategories) as $serviceCategoryId) {
+        $updateStatement->bindValue(':serviceCategoryId', $serviceCategoryId, \PDO::PARAM_INT);
+        $updateStatement->execute();
+
+        $selectStatement->bindValue(':serviceCategoryId', $serviceCategoryId, \PDO::PARAM_INT);
+        $selectStatement->execute();
+        $result = $selectStatement->fetch();
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:244]: You can use the new methods of CentreonDb now ;) Don't forget to catch the exception ;)

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 5]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+            action_type: ActionLog::ACTION_TYPE_ENABLE
+        );
     }
 }
 
-function disableServiceCategorieInDB(?int $sc_id = null, $sc_arr = [])
+function disableServiceCategorieInDB(?int $serviceCategoryId = null, $serviceCategories = [])
 {
-    if (!$sc_id && !count($sc_arr)) {
+    if (! $serviceCategoryId && ! count($serviceCategories)) {
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:256]:
```suggestion
    if (! $serviceCategoryId && empty($serviceCategories)) {
```

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 6]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+            object_type: ActionLog::OBJECT_TYPE_SERVICECATEGORIES,
+            object_id: $serviceCategoryId,
+            object_name: $result['sc_name'],
+            action_type: ActionLog::ACTION_TYPE_ENABLE
+        );
     }
 }
 
-function disableServiceCategorieInDB(?int $sc_id = null, $sc_arr = [])
+function disableServiceCategorieInDB(?int $serviceCategoryId = null, $serviceCategories = [])
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:254]:
```suggestion
function disableServiceCategorieInDB(?int $serviceCategoryId = null, array $serviceCategories = [])
```

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 7]
```diff
@@ -156,64 +167,141 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+            WHERE `sc_id` = :serviceCategoryId LIMIT 1
+        SQL
+    );
+    foreach (array_keys($serviceCategories) as $serviceCategoryId) {
+        $updateStatement->bindValue(':serviceCategoryId', $serviceCategoryId, \PDO::PARAM_INT);
+        $updateStatement->execute();
+
+        $selectStatement->bindValue(':serviceCategoryId', $serviceCategoryId, \PDO::PARAM_INT);
+        $selectStatement->execute();
+        $result = $selectStatement->fetch();
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:284]: You can use the new methods of CentreonDb now ;) Don't forget to catch the exception ;)

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 8]
```diff
@@ -308,18 +404,52 @@ function updateServiceCategorieInDB()
...
+    foreach (array_keys($serviceCategoryIds) as $serviceCategoryId) {
+        $serviceCategoryId = filter_var($serviceCategoryId, FILTER_VALIDATE_INT)
+            ?: throw new \Exception("Invalid service category id");
+
+        $selectStatement->bindValue(':serviceCategoryId', $serviceCategoryId, \PDO::PARAM_INT);
+        $selectStatement->execute();
+        $result = $selectStatement->fetch();
+
+        $deleteStatement->bindValue(':sc_id', $serviceCategoryId, \PDO::PARAM_INT);
+        $deleteStatement->execute();
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:446]: You can use the new methods of CentreonDb now ;) Don't forget to catch the exception ;)

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 9]
```diff
@@ -308,18 +404,52 @@ function updateServiceCategorieInDB()
...
 {
     global $pearDB, $centreon;
 
-    foreach (array_keys($ids) as $key) {
-        $scId = filter_var($key, FILTER_VALIDATE_INT);
-        $query = "DELETE FROM `service_categories` WHERE `sc_id` = :sc_id";
-        $statement = $pearDB->prepare($query);
-        $statement->bindValue(':sc_id', $scId, \PDO::PARAM_INT);
-        $statement->execute();
+    if (! $serviceCategoryIds) {
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:421]:
```suggestion
    if (is_null($serviceCategoryIds)) {
```

File: centreon/www/include/configuration/configObject/host/DB-Func.php [hunk 10]
```diff
@@ -297,7 +298,12 @@ function enableHostInDB($host_id = null, $host_arr = [])
         $hostName = $selectStatement->fetchColumn();
 
         signalConfigurationChange('host', (int) $hostId);
-        $centreon->CentreonLogAction->insertLog("host", $hostId, $hostName, "enable");
+        $centreon->CentreonLogAction->insertLog(
+            object_type: ActionLog::OBJECT_TYPE_HOST,
+            object_id: $hostId,
+            object_name: $hostName,
+            action_type: "enable"
```
Comment (a-launois) [inline centreon/www/include/configuration/configObject/host/DB-Func.php:305]:
```suggestion
            action_type: ActionLog::ACTION_TYPE_DISABLE
```

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 11]
```diff
@@ -156,64 +167,170 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+                object_type: ActionLog::OBJECT_TYPE_SERVICECATEGORIES,
+                object_id: $serviceCategoryId,
+                object_name: $result['sc_name'],
+                action_type: ActionLog::ACTION_TYPE_ENABLE
+            );
+        }
+    } catch(CentreonDbException $ex) {
+        CentreonLog::create()->error(
+            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+            message: $ex->getMessage(),
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:265]: Not only the message of the exception, it's not a good practice, you need context too.

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 12]
```diff
@@ -156,64 +167,170 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+                            object_type: ActionLog::OBJECT_TYPE_SERVICECATEGORIES,
+                            object_id: $maxId['maxid'],
+                            object_name: $sc_name,
+                            action_type: ActionLog::ACTION_TYPE_ADD,
+                            fields: $fields
+                        );
+                    } catch (CentreonDbException $ex) {
+                        CentreonLog::create()->error(
+                            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+                            message: $ex->getMessage(),
```
Comment (Drago37) [inline centreon/www/include/configuration/configObject/service_categories/DB-Func.php:209]: Not only the message of the exception, it's not a good practice, you need context too.

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 4 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: ok

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 7 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 8 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

**Review by Drago37 — CHANGES_REQUESTED** [review]

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 12 above)
Comment (jeremyjaouen) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php:209]: Can you provide an example of context

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 11 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php (hunk 12 above)
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 13]
```diff
@@ -156,64 +167,170 @@ function multipleServiceCategorieInDB($sc = [], $nbrDup = [])
...
+                object_type: ActionLog::OBJECT_TYPE_SERVICECATEGORIES,
+                object_id: $serviceCategoryId,
+                object_name: $result['sc_name'],
+                action_type: ActionLog::ACTION_TYPE_DISABLE
+            );
+        }
+    } catch (CentreonDbException $ex) {
+        CentreonLog::create()->error(
+            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+            message: $ex->getMessage(),
```
[the remark this answers is not in the document]
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

File: centreon/www/include/configuration/configObject/service_categories/DB-Func.php [hunk 14]
```diff
@@ -308,20 +433,62 @@ function updateServiceCategorieInDB()
...
+                object_id: $serviceCategoryId,
+                object_name: $result['sc_name'],
+                action_type: ActionLog::ACTION_TYPE_DELETE
+            );
+        }
+        $centreon->user->access->updateACL();
+    } catch(CentreonDbException $ex) {
+        CentreonLog::create()->error(
+            logTypeId: CentreonLog::TYPE_BUSINESS_LOG,
+            message: $ex->getMessage(),
```
[the remark this answers is not in the document]
Comment (Drago37) [reply centreon/www/include/configuration/configObject/service_categories/DB-Func.php]: good

**Review by Drago37 — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**